File: AGTIA_Objective-C/GooeySlideMenu/GooeySlideMenu/Classes/SlideMenuButton.m
//
//  SlideMenuButton.m
//  GooeySlideMenu
//
//  Created by fernando on 2017/2/4.
//  Copyright © 2017年 fernando. All rights reserved.
//

#import "SlideMenuButton.h"

@interface SlideMenuButton ()

@property (nonatomic, copy) NSString  *buttonTitle;

@end

@implementation SlideMenuButton

- (instancetype)initWithTitle:(NSString *)title {
    if (self = [super init]) {
        self.buttonTitle = title;
        self.buttonColor = [UIColor clearColor];
    }
    return self;
}

- (void)drawRect:(CGRect)rect {
    
    CGContextRef ctx = UIGraphicsGetCurrentContext();
    CGContextAddRect(ctx, rect);
    [self.buttonColor set];
    CGContextFillPath(ctx);
    
    UIBezierPath *roundedRactabglePath = [UIBezierPath bezierPathWithRoundedRect:CGRectInset(rect, 1, 1) cornerRadius:rect.size.height/2];
    [self.buttonColor setFill];
    [roundedRactabglePath fill];
    [[UIColor whiteColor] setStroke];
    roundedRactabglePath.lineWidth = 1;
    [roundedRactabglePath stroke];
    
    NSMutableParagraphStyle *paragraphStyle = [[NSParagraphStyle defaultParagraphStyle] mutableCopy];
    paragraphStyle.alignment = NSTextAlignmentCenter;
    NSDictionary *attr = @{NSParagraphStyleAttributeName:paragraphStyle, NSFontAttributeName:[UIFont systemFontOfSize:24], NSForegroundColorAttributeName:[UIColor whiteColor]};
    CGSize size = [self.buttonTitle sizeWithAttributes:attr];
    
    CGRect r = CGRectMake(rect.origin.x, rect.origin.y + (rect.size.height - size.height)/2.0, rect.size.width, size.height);
    [self.buttonTitle drawInRect:r withAttributes:attr];
}

- (void)touchesEnded:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    UITouch *touch = [touches anyObject];
    NSInteger tapCount = touch.tapCount;
    switch (tapCount) {
        case 1:
            if (self.buttonClickedBlock) {
                self.buttonClickedBlock();
            }
            break;
            
        default:
            break;
    }
}

@end

File: AGTIA_Swift/GooeySlideMenu/GooeySlideMenu/Classes/SlideMenuButton.swift
//
//  SlideMenuButton.swift
//  GooeySlideMenu
//
//  Created by fernando on 2017/2/5.
//  Copyright © 2017年 fernando. All rights reserved.
//

import UIKit

struct MenuButtonOptions {
    var title: String
    var buttonColor: UIColor
    var buttonClickBlock: ()->()
}

class SlideMenuButton: UIView {
    private var option: MenuButtonOptions
    
    init(option: MenuButtonOptions) {
        self.option = option
        super.init(frame: .zero)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func draw(_ rect: CGRect) {
        let context = UIGraphicsGetCurrentContext()
        context?.addRect(rect)
        option.buttonColor.set()
        context?.fillPath()
        
        let roundedRectanglePath = UIBezierPath(roundedRect: rect.insetBy(dx: 1, dy: 1), cornerRadius: rect.height / 2)
        option.buttonColor.setFill()
        roundedRectanglePath.fill()
        UIColor.white.setStroke()
        roundedRectanglePath.lineWidth = 1
        roundedRectanglePath.stroke()
        
        let paragraphStyle = NSParagraphStyle.default.mutableCopy() as! NSMutableParagraphStyle
        paragraphStyle.alignment = .center
        let attr = [NSParagraphStyleAttributeName: paragraphStyle, NSFontAttributeName: UIFont.systemFont(ofSize: 24), NSForegroundColorAttributeName: UIColor.white]
        let size = option.title.size(attributes: attr);
        
        let r = CGRect(x: rect.origin.x, y: rect.origin.y + (rect.size.height - size.height)/2.0, width: rect.size.width, height: size.height)
        option.title.draw(with: r, options: .usesLineFragmentOrigin, attributes: attr, context: nil)
    }
    
    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        let touch = touches.first
        if let touch = touch {
            let tapCount = touch.tapCount
            switch tapCount {
            case 1:
                option.buttonClickBlock()
            default:
                break
            }
        }
    }
}
